src/ngx_base_fetch.cc

```diff
@@ -138,9 +140,30 @@ void NgxBaseFetch::ReadCallback(const ps_event_data& data) {
   if (refcount == 0 || detached) {
     return;
   }
+
   ps_request_ctx_t* ctx = ps_get_request_context(r);
 
-  CHECK(data.sender == ctx->base_fetch);
+  // If our request context was zeroed, skip this event.
+  // See https://github.com/pagespeed/ngx_pagespeed/issues/1081
+  if (ctx == NULL) {
+    // Should not happen normally, when it does this message will cause our
+    // system tests to fail.
+    ngx_log_error(NGX_LOG_WARN, ngx_cycle->log, 0,
+        "pagespeed [%p] skipping event: request context gone", r);
+    return;
```

Comment:
This function processes events from pagespeed as we receive them through the named pipe, and subsequently drives nginx's request processing.  When the condition handled here arises, we have received an event originating from an `NgxBaseFetch` instance that is no longer associated to the request it originated from (the whole request context has been zeroed). 

Amongst other cases, nginx zeroes the active request context when it performs an internal redirect to serve a custom error page.

There was a flow/programming error earlier on the nginx side specific to .pagespeed. 404 responses which resolved a custom error page through such an internal redirect. I fixed that, and now I'm not able to reproduce this condition at all in any flow I can think of. 

So the changes for this function are about being more defensive against mistakes and making sure we fail the system tests on a bad request execution flow. For running production, my thought is that CHECK failing would take out the server, while not processing the event *might* take down a request. Processing an event that should not have been processed may make us output things we don't want, whereas not processing it when it should be may stall a request, which seems the better option.
We do log a warning about it, because when this happens we should take a look at it.